src/Microsoft.VisualStudio.Threading/JoinableTask`1.cs

```diff
@@ -47,10 +47,22 @@ internal JoinableTask(JoinableTaskFactory owner, bool synchronouslyBlocking, Joi
         /// </summary>
         /// <param name="cancellationToken">A cancellation token that will exit this method before the task is completed.</param>
         /// <returns>A task that completes after the asynchronous operation completes and the join is reverted, with the result of the operation.</returns>
-        public new async Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
+        public new Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
-            await base.JoinAsync(cancellationToken).ConfigureAwait(AwaitShouldCaptureSyncContext);
-            return await this.Task.ConfigureAwait(AwaitShouldCaptureSyncContext);
+            Task baseJoin = base.JoinAsync(cancellationToken);
+            if (baseJoin.IsCompleted && !cancellationToken.IsCancellationRequested)
```

Comment:
📝 If there are still concerns here, we could change this to:

```suggestion
            if (baseJoin.Status == TaskStatus.RanToCompletion)
```

Reply:
I'm curious why you added the `&& !cancellationToken.IsCancellationRequested` to your condition. Why would you want a canceled request to then take the slow path, even if the task has already completed?

Reply:
I'm removing the oddity and simplifying the optimization so it doesn't even call `base.JoinAsync` when `this.IsCompleted`.